File: src/algorithm/Tree/ZigzagLevelOrder103.java
package algorithm.Tree;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * 103. 二叉树的锯齿形层序遍历
 * 给定一个二叉树，返回其节点值的锯齿形层序遍历。
 * （即先从左往右，再从右往左进行下一层遍历，以此类推，层与层之间交替进行）。
 * 例如：
 * 给定二叉树 [3,9,20,null,null,15,7],
 *     3
 *    / \
 *   9  20
 *     /  \
 *    15   7
 * 返回锯齿形层序遍历如下：
 * [
 *   [3],
 *   [20,9],
 *   [15,7]
 * ]
 */
public class ZigzagLevelOrder103 {

    /**
     * 这题是102--层序遍历的变形（bfs遍历二叉树）
     * 用到层序遍历解题的：102层序遍历、103二叉树的锯齿形层序遍历、637二叉树的层平均值
     * 104二叉树的最大深度也可以使用层遍历、111二叉树的最小深度
     */


    /**
     * 这题不应该考虑如何将节点按照奇偶层顺序插入到队列当中，这样会导致代码繁琐
     * 而是遍历任为层序遍历但是将根据奇偶层将结果（节点值）头插或尾插如tmp链表当中
     */


    /**
     * 更为高效的思路是结点仍然按照中序遍历的顺序加入到结点队列当中
     * 然后按照当前结点属于奇数层还是偶数层来头插或尾插进结果集合
     * 具体的：
     * 如果当前处于奇数层，我们每次将被遍历到的元素插入至双端队列的末尾。
     * 如果当前处于偶数层，我们每次将被遍历到的元素插入至双端队列的头部。
     */
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res=new ArrayList<>();
        if(root==null){
            return res;
        }
        //奇数层栈
        LinkedList<TreeNode> list=new LinkedList<>();
        list.add(root);
        boolean isOdd=true;
        while(!list.isEmpty()){
            List<Integer> tmp=new LinkedList<>();
            int n=list.size();
            for(int i=0;i<n;++i) {
                TreeNode node = list.poll();
                if (isOdd) {
                    //奇数层，数据尾插
                    tmp.add(node.val);
                } else {
                    //偶数层，数据头插
                    tmp.add(0,node.val);
                }
                if(node.left!=null){
                    list.add(node.left);
                }
                if(node.right!=null){
                    list.add(node.right);
                }
            }
            res.add(tmp);
            isOdd=!isOdd;
        }
        return res;
    }

    //dfs版，虽然是深度优先遍历，但是依旧按照先左子节点，后右子节顺序（加之level代表层数）
    public List<List<Integer>> zigzagLevelOrder2(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        traversal(root, res, 0);
        return res;
    }

    //level对应的就是结果集合的索引
    private void traversal(TreeNode root, List<List<Integer>> res, int level) {
        if (root == null) {
            return;
        }
        if (res.size() == level) {
            res.add(new ArrayList<Integer>());
        }
        //相较层序遍历dfs版只有这里的不同
        if ((level & 1) == 1){
            //奇数层头插
            res.get(level).add(0, root.val);
        } else {
            res.get(level).add(root.val);
        }
        traversal(root.left, res, level + 1);
        traversal(root.right, res, level + 1);
    }


    /**
     * 但队列---但是是按奇偶层模拟了节点如何插入到队列中
     */
    public List<List<Integer>> zigzagLevelOrder1(TreeNode root) {
        List<List<Integer>> res=new ArrayList<>();
        if(root==null) return res;
        Deque<TreeNode> queue=new LinkedList<>();
        queue.add(root);
        boolean isEven=true;
        while(!queue.isEmpty()){
            int size=queue.size();
            List<Integer> tmp=new ArrayList<>();
            for(int i=0;i<size;++i){
                TreeNode curr;
                if(isEven){
                    curr=queue.pollLast();
                    if(curr.left!=null){
                        queue.addFirst(curr.left);
                    }
                    if(curr.right!=null){
                        queue.addFirst(curr.right);
                    }
                }else{
                    curr=queue.pollFirst();
                    if(curr.right!=null){
                        queue.addLast(curr.right);
                    }
                    if(curr.left!=null){
                        queue.addLast(curr.left);
                    }
                }
                tmp.add(curr.val);
            }
            res.add(tmp);
            isEven=!isEven;
        }
        return res;
    }

    /**
     * 自己写的双栈实现
     * 这个的实现是改变结点插入到双向链表的顺序（并非直接的层序遍历顺序）,不推荐该思路
     */
    public List<List<Integer>> zigzagLevelOrderMy(TreeNode root) {
        List<List<Integer>> res=new ArrayList<>();
        if(root==null){
            return res;
        }
        //奇数层栈
        LinkedList<TreeNode> odd=new LinkedList<>();
        //偶数层栈
        LinkedList<TreeNode> even=new LinkedList<>();
        odd.add(root);
        boolean isOdd=true;
        while(!odd.isEmpty()||!even.isEmpty()){
            List<Integer> tmp=new ArrayList<>();
            //根据是奇数层还是偶数层来改变加入结点栈的元素顺序
            if(isOdd){
                int size=odd.size();
                for(int i=0;i<size;++i) {
                    TreeNode node=odd.pollFirst();
                    tmp.add(node.val);
                    if(node.left!=null){
                        even.addFirst(node.left);
                    }
                    if(node.right!=null){
                        even.addFirst(node.right);
                    }
                }
            }else {
                int size=even.size();
                for(int i=0;i<size;++i) {
                    TreeNode node=even.pollFirst();
                    tmp.add(node.val);
                    if(node.right!=null){
                        odd.addFirst(node.right);
                    }
                    if(node.left!=null){
                        odd.addFirst(node.left);
                    }
                }
            }
            res.add(tmp);
            isOdd=!isOdd;
        }
        return res;
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}

File: go/src/Tree/ZigzagLevelOrder103.go
package main

func zigzagLevelOrder(root *TreeNode) [][]int {
	res := make([][]int, 0)
	if root == nil {
		return res
	}
	q := []*TreeNode{root}
	level := 1
	for len(q) > 0 {
		tmp, size := []int{}, len(q)
		for i := 0; i < size; i++ {
			p := q[0]
			q = q[1:]
			tmp = append(tmp, p.Val)
			if p.Left != nil {
				q = append(q, p.Left)
			}
			if p.Right != nil {
				q = append(q, p.Right)
			}
		}
		if (level & 1) == 0 {
			for i, j := 0, len(tmp)-1; i < j; {
				tmp[i], tmp[j] = tmp[j], tmp[i]
				i++
				j--
			}
		}
		res = append(res, tmp)
		level++
	}
	return res
}

// 另外一种写法（更偏GO习惯）
func zigzagLevelOrderII(root *TreeNode) (ans [][]int) {
	// 第一点不同就是返回值直接在函数中定义好名字可以直接return
	if root == nil {
		return
	}
	queue := []*TreeNode{root}
	for level := 0; len(queue) > 0; level++ {
		vals := []int{}
		//将切片赋值，可以直接对赋值后的切片进行range操作
		q := queue
		//原队列赋空值
		queue = nil
		for _, node := range q {
			vals = append(vals, node.Val)
			//但是添加元素的操作还是要对原队列进行的
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		// 本质上和层序遍历一样，我们只需要把奇数层的元素翻转即可
		if level%2 == 1 {
			for i, n := 0, len(vals); i < n/2; i++ {
				vals[i], vals[n-1-i] = vals[n-1-i], vals[i]
			}
		}
		ans = append(ans, vals)
	}
	return
}

// 迭代的第三种实现，直接add元素。内存效率比前两个低，同理dfs也可以同样套路
func zigzagLevelOrder3(root *TreeNode) (res [][]int) {
	if root == nil {
		return
	}
	queue := []*TreeNode{root}
	left2Right := true
	for ; len(queue) != 0; left2Right = !left2Right {
		item := []int{}
		for size := len(queue); size > 0; size-- {
			pop := queue[0]
			queue = queue[1:]
			if left2Right {
				item = append(item, pop.Val)
			} else {
				//不同点，切片头插的不同实现
				//这里是直接重建一个[]int切片，内存效率低
				item = append([]int{pop.Val}, item...)
			}
			if pop.Left != nil {
				queue = append(queue, pop.Left)
			}
			if pop.Right != nil {
				queue = append(queue, pop.Right)
			}
		}
		res = append(res, item)
	}
	return
}

// dfs实现
func zigzagLevelOrder4(root *TreeNode) (res [][]int) {
	if root == nil {
		return
	}
	//先申明了这个方法的函数类型再对方法句柄赋值,才能进行方法自递归调用
	var dfs func(root *TreeNode, depth int)
	dfs = func(root *TreeNode, depth int) {
		if root == nil {
			return
		}
		if depth+1 > len(res) {
			res = append(res, []int{})
		}
		if depth&1 == 0 {
			res[depth] = append(res[depth], root.Val)
		} else {
			res[depth] = append([]int{root.Val}, res[depth]...)
		}
		dfs(root.Left, depth+1)
		dfs(root.Right, depth+1)
	}

	dfs(root, 0)
	return
}
